### ISSUE TITLE
Tcomb breaks object type checking

### ISSUE THREAD

**Issue by ivan-kleshnin** [gcanti/tcomb#142]
I'm toying with **Tcomb** last days and did find that it wraps objects in its own structures.

This is a dealbreaker because noone wants another OOP. 
There is no reason to go further than just freezing an Object.

```js
function Struct(value, path) {
  // !!! evil
  if (Struct.is(value)) { // makes Struct idempotent
    return value;
  }

  ...

  // !!! evil
  if (!isInstanceOf(this, Struct)) { // makes `new` optional
    return new Struct(value);
  }

  ... 

  // this is ok
  if (process.env.NODE_ENV !== 'production') {
    Object.freeze(this);
  }
}
```

I understand you made it for performance, but see:

```js
import isPlainObject from "is-plain-object";

// declare Robot

console.log(isPlainObject(Robot({
  id: "ebffa860-5660-4561-a373-288f33544b71",
  name: "Alec",
  manufacturer: "USA",
  assemblyDate: new Date(),
}))); // false
````

Object checking is no longer working, cause Object is wrapped.
This cuts down everything.

I'm skeptical even about that performance gain because it seems most values will be passed through typing system only once. But this is another story. 

Object wrapping is OOP and requires from all code to be Tcomb-aware which goes straight against the purpose of the lib.

`Something.is` can't replace native JS. 
It may be supplemental, but all values must stay unwrapped to be compatible with existing ecosystem.
By wrapping natives you gain better performance and fashionable `console.log` but lose ecosystem compatibility. 100% not worth it.

**Comment by ivan-kleshnin** [comment]
I see `Array.isArray` is working. So something in-between going here.

Let's dig. Both `console.log` have to return **true** to keep Tcomb compatible.
```
import isPlainObject from "is-plain-object";
import Tc from "tcomb";

let MyObj = Tc.struct({name: Tc.String});
let MyList = Tc.list(Tc.Number);

let myObj = MyObj({name: "jac"});
let myList = MyList([1, 2, 3]);

console.log(isPlainObject(myObj)); // false
console.log(Array.isArray(myList)); // true
```

**Comment by ivan-kleshnin** [comment]
Exactly this check is failing now:

```js
function isPlainObject(o) {
  ...
  // If constructor does not have an Object-specific method
  if (prot.hasOwnProperty('isPrototypeOf') === false) {
    return false;
  }
}
```

**Comment by gcanti** [comment]
Hi!
all tcomb types are identity functions except the structs which are normal JavaScript constructors:

```js
// instead of...
function Person(value) {
  // manual type checking here...
  this.name = value.name;
  this.surname = value.surname
}

// ...you can write
var Person = t.struct({
  name: t.String,
  surname: t.String
});
```

**Comment by gcanti** [comment]
`isPlainObject` strikes back :-D

(I did find problems also with redux, actions **must** be plain objects)

In general I don't understand the reason of imposing `Object` as a constructor.

**Comment by ivan-kleshnin** [comment]
Sorry, I didn't quite understood what are you going to do about this.
We need to get `User(x) == x` irrelevant to what is `User` exactly.
Can we make `structs` identity functions as well?

**Comment by gcanti** [comment]
> We need to get User(x) == x irrelevant to what is User.

For the built-in combinators a weaker property holds (https://github.com/gcanti/tcomb/blob/master/GUIDE.md#the-idea):

2. the type `T` is idempotent, that is `T(T(value)) = T(value)`

With the only exception of the struct combinator, for all the remaining combinators your property holds.

> Can we make structs identity functions as well?

No, it can't be done without breaking changes. 
Morevoer I'm quite happy with the current implementation:

- fast `is(x)` function
- a way to attach functions to the prototype

As I stated above I don't understand the reason of imposing `Object` as a constructor
but I'm open to discussions, maybe I'm missing something.

Anyway you could define your own combinator `plain(props, name)` such that

```
T = plain(props, name);
T(value) === value for all values
```

**Comment by gcanti** [comment]
> a way to attach functions to the prototype

Just a remark on that. I tend to see my code as 

1. a bunch of immutable data structures
2. a bunch of functions acting on them
3. some mutable classes (ideally just one) holding the state
4. preference of composition over inheritance (the `extend` function on structs is just props composition)

1 + 2 together are what I consider the *domain model* of my application.

Now strictly speaking you don't **need** prototypes but they comes to handy

```js
const Person = t.struct({
  name: t.String,
  surname: t.String
});

// instead of..
function getFullName(person) {
  return `${person.name} ${person.surname}`;
}

// ...you COULD write
Person.prototype.getFullName = function () {
  return `${this.name} ${this.surname}`;
};
```

and then be sure the functions related to `Person` are accessible everytime you are managing a `Person` instance.

**Comment by ivan-kleshnin** [comment]
> As I stated above I don't understand the reason of imposing Object as a constructor

I don't understand this sentence.

> a bunch of immutable data structures
> a bunch of functions acting on them

This is cool, but all general purpose functions should work with Tcomb, right?
Now `isPlainObject` check is broken.

**Comment by ivan-kleshnin** [comment]
Maybe I should give an additional context because I feel we don't quite understand each other.

See this gist:
https://gist.github.com/ivan-kleshnin/1aac5a88eabb7388a802

This is a Tcomb-based parser which I'm going to use to convert form data to business data.
There is a check `isPlainObject` which can't be replaced without weird hacky assumptions.
And it does not work now.

**Comment by ivan-kleshnin** [comment]
// ...you COULD write
```js
Person.prototype.getFullName = function () {
  return `${this.name} ${this.surname}`;
};
```

> and then be sure the functions related to Person are accessible everytime you are managing a Person instance.

This is OOP which I avoid at all costs. I can get the same functionality by just importing standalone function which is easier to test and I'm free of prototype mess.

**Comment by ivan-kleshnin** [comment]
My options are:

1. wrap Tcomb, replacing `Tcomb.struct` with something that is compatible with `isPlainObject`. It will probably require to disable `freeze` and fake `isPrototypeOf` function.
2. wrap `isPlainObject` to make it compatible with Tcomb.

First option is way safer.

**Comment by gcanti** [comment]
> What do you mean by "imposing Object as a constructor"?

From the README of is-plain-object:

> Returns true if an object was created by the `Object` constructor.

`isPlainObject` is not "broken", it just returns `false` for every object obtained from another constructor

```js
// vanilla JavaScript constructor
function Person(name) {
  this.name = name;
}

var myobject = new Person('Giulio');
console.log(isPlainObject(myobject)); // => false, but myobject IS an object
```

If a library requires objects such that `isPlainObject(myobject) === true` this means that the library is banning all JavaScript constructors and I don't see a compelling reason to do that. It seems to me an unnecessary constraint. Since you are using `isPlainObject` could you please tell me the value of such a ban? Maybe you can suggest me a good reason and I can change my mind.

> This is OOP which I avoid at all costs

```js
Person.prototype.getFullName = function () {
  return `${this.name} ${this.surname}`;
};
```

It's just an implementation detail, from a math point of view is just a function.

> I prefer to use standalone functions which are easier to test and I'm free of prototype mess

Me too. Using the prototype is just an *option*.

**Comment by giogonzo** [comment]
Seems to me that `isPlainObject` is generally misused  as a guarantee that a value is serializable (`isPlainObject` is weaker than, actually has nothing to do with, `value <eq> JSON.parse(JSON.stringify(value))`)

A "serializabilty check" is in general unfeasible for performance reasons, it should be something all libs can do only when `process.env.NODE_ENV === 'development'`

@ivan-kleshnin also see https://github.com/rackt/redux/issues/437

@gcanti one single thing it's actually *weird* about all this is:

> ```js
console.log(isPlainObject(myStruct)); // false
console.log(Array.isArray(myList)); // true
```

Using a Symbol (similarly as suggested [here](https://github.com/rackt/redux/issues/437#issuecomment-129374459)) to hold the Struct constructor on POJOs might be a viable solution, still don't think it could be done without breaking changes

Answer should probably (same as issue linked above): it's something can easily be done in user-land (`POJO` combinator)

**Comment by ivan-kleshnin** [comment]
> If a library requires objects such that isPlainObject(myobject) === true this means that the library is banning all JavaScript constructors and I don't see a compelling reason to do that. It seems to me an unnecessary constraint. Since you are using isPlainObject could you please tell me the value of such a ban? Maybe you can suggest me a good reason and I can change my mind.

Let's discuss.

The JavaScript has one quirk, comparing to most other languages. 
We use `Object` in the role of dicts or records. The problem is that `Object` is a prototype of all other types, including `Array`, `Date`, etc. There is a new `Map` proposal but it's still in draft and it's very inconvenient currently (console.log does not reveal it's structure, it's longer to write, etc...).

So we use `{}` as dict and need to apply typechecking to differ that "plain" Objects from other objects.
In case of plain objects it's often required to traverse them to leaves while in case of other objects like `Date`, `Error` and `RegExp` and (some?) custom user objects this traversal must not be applied.

The limitation is that we have zero knowledge about uknown (custom) types. JavaScript has no interfaces. No protocols. Not even established conventions of what is "array-like" or "dict-like". 

So we basically have only two possible solutions.
1. Treat all unknown (custom) types like `Object`. I.e. like traversable trees of atrributes.
2. Treat all unknown (custom) types like immutable untraversable values.

Option 1. is a disaster. Think of `Date` wrapped in `Moment`. Or `ReactNode` object. If we're trying to traverse them because they are `not Array` we're screwed.

So it's not a "ban" of JS Constructors. It's one of the two only possible logic treatments of them.

**Comment by ivan-kleshnin** [comment]
Actually I agree with @gcanti.

**is-plain-object** is a crappy library. It relies on one external library and use **four** types of is-object checking with blurry and random semantics. Just insane.

Here's all code I need:

```js
function isPlainObject(o) {
  return toString.call(o) === '[object Object]';
}

console.log(isPlainObject(new Object())); // true
console.log(isPlainObject({})); // true
console.log(isPlainObject((new Tc.struct({name: Tc.String}))({name: "jac"}))); // true

console.log(isPlainObject(new Date())); // false
console.log(isPlainObject(new Error())); // false
console.log(isPlainObject(new RegExp())); // false
```

Though my above reasoning still applies, my current solution will be to get rid of this library.
